File: Unit - 2/Basic Traversal and Search Techniques/Java Implementation/BFS.java
//Implement BFS algorithm on a graph represented as Adjacency matrix
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class BFS {
    private static int count = 0;
    private static Queue<Integer> queue = new LinkedList<>();

    private static void BFS(int[][]graph, int[] visitedNodes){
        for(int v = 0; v < visitedNodes.length; v++){
            if(visitedNodes[v] == 0){
                bfs(v, graph, visitedNodes);
            }
        }
    }

    private static void bfs(int v, int[][]graph, int[] visitedNodes){
        count++;
        visitedNodes[v] = count;
        queue.add(v);
        while(!queue.isEmpty()){
            int frontElement = queue.peek();
            System.out.print(frontElement + " ");
            for(int w = 0; w < visitedNodes.length; w++){
                if(visitedNodes[w] == 0 && graph[frontElement][w] == 1){
                    count++;
                    visitedNodes[w] = count;
                    queue.add(w);
                }
            }
            queue.remove();
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of nodes: ");
        int nodeCount = sc.nextInt();
        int[][] graph = new int[nodeCount][nodeCount];
        int[] visitedNodes = new int[nodeCount];
        System.out.println("Enter the adjacency matrix: ");
        for(int i = 0; i < nodeCount; i++){
            for(int j = 0; j < nodeCount; j++){
                graph[i][j] = sc.nextInt();
            }
        }
        System.out.println("\nThe order in which the nodes are visited is:");
        BFS(graph, visitedNodes);
    }
}

File: Assignment programs/BFS.c
//Implement BFS algorithm on a graph represented as Adjacency matrix
#include<stdio.h>
#define MAX 100

int count = 0, queue[MAX], front = 0, rear = -1;
void insertQueue(int element)
{
    queue[++rear] = element;
}

void deleteQueue()
{
    queue[front++];
    if(front > rear)
    {
        front = 0;
        rear = -1;
    }
}

void readMatrix(int matrix[][MAX], int n)
{
    int i, j;
    for(i = 0; i < n; i++)
    {
        for(j = 0; j < n; j++)
        {
            scanf("%d", &matrix[i][j]);
        }
    }
}

void markUnvisited(int visitedNodes[], int nodeCount)
{
    int i;
    for(i = 0; i < nodeCount; i++)
    {
        visitedNodes[i] = 0;
    }
}

void bfs(int v, int graph[][MAX], int visitedNodes[], int nodeCount)
{
    count++;
    visitedNodes[v] = count;
    insertQueue(v);
    while(front <= rear)
    {
        int frontElement = queue[front], w;
        printf("%d ", frontElement);
        for(w = 0; w < nodeCount; w++)
        {
            if(graph[frontElement][w] == 1 && visitedNodes[w] == 0)
            {
                count++;
                visitedNodes[w] = count;
                insertQueue(w);
            }
        }
        deleteQueue();
    }
}

void BFS(int graph[][MAX], int visitedNodes[], int nodeCount)
{
    int v;
    for(v = 0; v < nodeCount; v++)
    {
        if(visitedNodes[v] == 0)
        {
            bfs(v, graph, visitedNodes, nodeCount);
        }
    }
}

int main()
{
    int nodeCount, graph[MAX][MAX], visitedNodes[MAX];
    printf("Enter the number of nodes: ");
    scanf("%d", &nodeCount);
    printf("Enter the adjacency matrix:\n");
    readMatrix(graph, nodeCount);
    markUnvisited(visitedNodes, nodeCount);
    printf("The order in which the nodes are visited is:\n");
    BFS(graph, visitedNodes, nodeCount);
    printf("\n");
    return 0;
}